libfreerdp/core/peer.c

```diff
@@ -758,44 +779,62 @@ static int peer_recv_callback(rdpTransport* transport, wStream* s, void* extra)
 				 * During reactivation sequence the client might sent some input or channel data
 				 * before receiving the Deactivate All PDU. We need to process them as usual.
 				 */
-				if (peer_recv_pdu(client, s) < 0)
+				ret = peer_recv_pdu(client, s);
+				if (ret < 0)
 				{
 					WLog_ERR(TAG,
 					         "%s: %s - "
 					         "peer_recv_pdu() fail",
 					         __FUNCTION__, rdp_get_state_string(rdp));
-					return -1;
 				}
 			}
 
 			break;
 
 		case CONNECTION_STATE_FINALIZATION:
-			if (peer_recv_pdu(client, s) < 0)
+			ret = peer_recv_pdu(client, s);
+			if (ret < 0)
 			{
 				WLog_ERR(TAG, "%s: %s - peer_recv_pdu() fail", __FUNCTION__,
 				         rdp_get_state_string(rdp));
-				return -1;
 			}
-
 			break;
 
 		case CONNECTION_STATE_ACTIVE:
-			if (peer_recv_pdu(client, s) < 0)
+			ret = peer_recv_pdu(client, s);
+			if (ret < 0)
 			{
 				WLog_ERR(TAG, "%s: %s - peer_recv_pdu() fail", __FUNCTION__,
 				         rdp_get_state_string(rdp));
-				return -1;
 			}
 
 			break;
 
 		default:
 			WLog_ERR(TAG, "%s state %d", rdp_get_state_string(rdp), rdp_get_state(rdp));
-			return -1;
+			break;
 	}
 
-	return 0;
+	return ret;
+}
+
+static int peer_recv_callback(rdpTransport* transport, wStream* s, void* extra)
+{
+	int rc = 0;
+	do
+	{
+		switch (rc)
+		{
+			default:
+				rc = peer_recv_callback_internal(transport, s, extra);
+				break;
+			case 2:
```

Comment:
Not 100% sure but when I look at the following code in transport.c

```
		recv_status = transport->ReceiveCallback(transport, received, transport->ReceiveExtra);
		...
		/* session redirection or activation */
		if (recv_status == 1 || recv_status == 2)
```

this looping/rc handling might cause problems?

Reply:
client side code, the other one is server side